controller/rancher/rancher.go

```diff
@@ -488,6 +550,238 @@ func (fetcher *RCertificateFetcher) UpdateEndpoints(lbSvc *metadata.Service, eps
+	fetcher.mu.RLock()
+	currentDefCertD := fetcher.DefaultCertDir
+	fetcher.mu.RUnlock()
+
+	if currentDefCertD != defaultCertDir {
+		//defaultCertDir value changed, read it again.
+		fetcher.Watcher.Remove(defaultCertDir)
+		fetcher.mu.Lock()
+		fetcher.TempCertsMap = make(map[string]*config.Certificate)
+		err := filepath.Walk(defaultCertDir, fetcher.readCertificate)
+		if err != nil {
+			fetcher.DefaultCertDir = ""
+			fetcher.DefaultCert = nil
+			logrus.Errorf("Error %v reading defaultCertDir  %v", err, defaultCertDir)
+		} else {
+			fetcher.DefaultCertDir = defaultCertDir
+			for _, cert := range fetcher.TempCertsMap {
+				fetcher.DefaultCert = cert
+			}
+			fetcher.Watcher.Add(defaultCertDir)
+		}
+		fetcher.mu.Unlock()
+	}
+	return fetcher.DefaultCert, nil
+}
+
+func (fetcher *RCertificateFetcher) readCertificate(path string, f os.FileInfo, err error) error {
+	if f != nil && f.IsDir() {
+		logrus.Debugf("walking %v", path)
+		isCertFound := false
+		cert := config.Certificate{}
+		cert.Name = f.Name()
+		files, err := ioutil.ReadDir(path)
+		if err != nil {
+			return err
+		}
+		for _, file := range files {
+			if !file.IsDir() {
+				contentBytes, err := fetcher.evaluatueLinkAndReadFile(path, file.Name())
+				if err != nil {
+					return err
+				}
+				if file.Name() == "fullchain.pem" {
+					isCertFound = true
+					cert.Cert = string(*contentBytes)
+				} else if file.Name() == "privkey.pem" {
+					isCertFound = true
+					cert.Key = string(*contentBytes)
+				}
+			}
+		}
+		if isCertFound {
+			fetcher.TempCertsMap[path] = &cert
+		}
+	}
+	return nil
+}
+
+func (fetcher *RCertificateFetcher) evaluatueLinkAndReadFile(relativePath string, fileName string) (*[]byte, error) {
+	filePath := path.Join(relativePath, fileName)
+
+	absFilePath, err := filepath.Abs(filePath)
+	if err != nil {
+		logrus.Errorf("Error forming path to file %s, error: %v", filePath, err)
+		return nil, err
+	}
+
+	fInfo, err := os.Lstat(absFilePath)
+	if os.IsNotExist(err) {
+		logrus.Errorf("File %s does not exist", absFilePath)
+		return nil, err
+	}
+
+	targetPath := absFilePath
+	if fInfo.Mode()&os.ModeSymlink != 0 {
+		//read symlink
+		targetPath, err := filepath.EvalSymlinks(absFilePath)
+		if err != nil {
+			logrus.Debugf("File %s pointed by symlink %s does not exist", targetPath, absFilePath)
+			return nil, err
+		}
+	}
+	//read target file
+	return fetcher.readFile(targetPath)
+}
+
+func (fetcher *RCertificateFetcher) readFile(filePath string) (*[]byte, error) {
+	contentBytes, err := ioutil.ReadFile(filePath)
+	if err != nil {
+		logrus.Errorf("Error reading file %s, error: %v", filePath, err)
+		return nil, err
+	}
+	return &contentBytes, nil
+}
+
+func (fetcher *RCertificateFetcher) addWatches() {
+	if fetcher.CertDir != "" {
+		err := fetcher.Watcher.Add(fetcher.CertDir)
+		if err != nil {
+			logrus.Errorf("Failed to add watch for Dir %v", fetcher.CertDir)
```

Comment:
I think the only possible error condition is file path not found, however since we add watch after reading file on that path, we are less likely to hit an error.

Other option we have is on error, return error to the whole "readCerts" method which will error on the haproxy applyconfig - does that sound ok?